bench_test.go

```diff
@@ -4,80 +4,121 @@
+	benchtime := flag.Lookup("test.benchtime").Value.String()
+	if b.N == 1 && strings.HasSuffix(benchtime, "x") && benchtime != "1x" {
+		return
+	}
+
 	garbleBin := filepath.Join(b.TempDir(), "garble")
 	if runtime.GOOS == "windows" {
 		garbleBin += ".exe"
 	}
+	err := exec.Command("go", "build", "-o="+garbleBin).Run()
+	qt.Assert(b, err, qt.IsNil)
+
+	// We collect extra metrics.
+	var userTime, systemTime int64
+
+	var cachedTime int64
 
-	if err := exec.Command("go", "build", "-o="+garbleBin).Run(); err != nil {
-		b.Fatalf("building garble: %v", err)
+	outputBin := filepath.Join(b.TempDir(), "output")
+	sourceDir := filepath.Join(b.TempDir(), "src")
+	err = os.Mkdir(sourceDir, 0o777)
+	qt.Assert(b, err, qt.IsNil)
+
+	writeSourceFile := func(name string, content []byte) {
+		err := os.WriteFile(filepath.Join(sourceDir, name), content, 0o666)
+		qt.Assert(b, err, qt.IsNil)
 	}
+	writeSourceFile("go.mod", []byte("module test/main"))
+	writeSourceFile("main.go", benchSourceMain)
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		// First we do a fresh build, using a new GOCACHE.
+		// and the second does an incremental rebuild reusing the cache.
+		gocache, err := os.MkdirTemp(b.TempDir(), "gocache-*")
+		qt.Assert(b, err, qt.IsNil)
+		env := append(os.Environ(), "GOGARBLE=*", "GOCACHE="+gocache)
+		args := []string{"build", "-v", "-o=" + outputBin, sourceDir}
+
+		cmd := exec.Command(garbleBin, args...)
+		cmd.Env = env
+		cmd.Dir = sourceDir
+
+		out, err := cmd.CombinedOutput()
+		qt.Assert(b, err, qt.IsNil, qt.Commentf("output: %s", out))
+		// Ensure that we built all packages, as expected.
+		qt.Assert(b, rxBuiltRuntime.Match(out), qt.IsTrue)
+		qt.Assert(b, rxBuiltMain.Match(out), qt.IsTrue)
 
-	for _, name := range [...]string{"Cache", "NoCache"} {
-		b.Run(name, func(b *testing.B) {
-			buildArgs := []string{"build", "-o=" + b.TempDir()}
-			switch name {
-			case "Cache":
-				buildArgs = append(buildArgs, "./testdata/bench-cache")
-
-				// Ensure the build cache is warm,
-				// for the sake of consistent results.
-				cmd := exec.Command(garbleBin, buildArgs...)
-				if out, err := cmd.CombinedOutput(); err != nil {
-					b.Fatalf("%v: %s", err, out)
-				}
-			case "NoCache":
-				buildArgs = append(buildArgs, "./testdata/bench-nocache")
-			default:
-				b.Fatalf("unknown name: %q", name)
-			}
-
-			// We collect extra metrics.
-			var userTime, systemTime int64
-
-			b.ResetTimer()
-			b.RunParallel(func(pb *testing.PB) {
-				for pb.Next() {
-					cmd := exec.Command(garbleBin, buildArgs...)
-					if name == "NoCache" {
-						gocache, err := os.MkdirTemp(b.TempDir(), "gocache-*")
-						if err != nil {
-							b.Fatal(err)
-						}
-						cmd.Env = append(os.Environ(), "GOCACHE="+gocache)
-					}
-					if out, err := cmd.CombinedOutput(); err != nil {
-						b.Fatalf("%v: %s", err, out)
-					}
-
-					userTime += int64(cmd.ProcessState.UserTime())
-					systemTime += int64(cmd.ProcessState.SystemTime())
```

Comment:
Oops, this was racy as heck, too. Multiple goroutines (from the parallel benchmark) adding to a shared integer without atomics nor a lock.